fix #1935, add topological sort algorithm for graph with circles.

pangolulu opened the pull request.

clockfly commented:
Thanks! @pangolulu

clockfly commented:
Can you add the algorithm link you use in the source code.
Can you add some unit test for it? The test is in file GraphSpec.

pangolulu commented:
The algorithm is following [http://www.drdobbs.com/database/topological-sorting/184410262](). @clockfly

clockfly commented on core/src/main/scala/io/gearpump/util/Graph.scala:
```diff
@@ -254,6 +254,90 @@ class Graph[N, E](vertexList: List[N], edgeList: List[(N, E, N)]) extends Serial
+  def findCircles: mutable.Set[mutable.Set[N]] = {
+    val inStack = mutable.Map[N, Boolean]()
+    val stack = mutable.Stack[N]()
+    val DFN = mutable.Map[N, Int]()
```
DFN, LOW is not good varible name. Maybe index and lowlink better? 
https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm

clockfly commented on core/src/main/scala/io/gearpump/util/Graph.scala:
```diff
@@ -254,6 +254,90 @@ class Graph[N, E](vertexList: List[N], edgeList: List[(N, E, N)]) extends Serial
+   * Return an iterator of vertex in topological order of graph with circles
+   * The node returned by Iterator is stable sorted.
+   */
+  def topologicalOrderWithCirclesIterator: Iterator[N] = {
```
Are you sure it is "stable sorted"?

Since you are using Set, and Set is not stable when expanded.

clockfly commented on core/src/test/scala/io/gearpump/util/GraphSpec.scala:
```diff
@@ -178,6 +178,30 @@ class GraphSpec extends PropSpec with PropertyChecks with Matchers {
+  }
+
+  property("Topological sort of graph with circles should work properly") {
+    val graph = Graph(1 ~> 3 ~> 4 ~> 6 ~> 5, 4 ~> 1, 1 ~> 2 ~> 4)
```
Can you add a test case which have two strong connected components, and one component ranks over the other by topology order. 

So that we can examine the behavior for multiple cycles.

clockfly commented on core/src/main/scala/io/gearpump/util/Graph.scala:
```diff
@@ -254,6 +254,90 @@ class Graph[N, E](vertexList: List[N], edgeList: List[(N, E, N)]) extends Serial
   /**
+   * Return all circles in graph.
+   */
+  def findCircles: mutable.Set[mutable.Set[N]] = {
```
Is the findCircles used elsewhere? if not, probably make it private

clockfly replied:
Please add the algorithm link as comments in source code.

clockfly commented:
How about making this change, change the reference of topologicalOrderIterator to use topologicalOrderWithCirclesIterator instead, and see whether the SAMOA module works?

pangolulu commented:
@clockfly 
I have changed the variables name, and make the sort stable using ``MutableList``.

I changed the GraphSpec to test whether ``topologicalOrderWithCirclesIterator`` equals ``topologicalOrderIterator`` in the graph with no circle but contains more than one sort to see if the actual sort is stable.

And test ``topologicalOrderIterator`` in the complex graph with more than one circles.

pangolulu commented:
@clockfly 
Yes, I will try ``topologicalOrderWithCirclesIterator `` in SAMOA.

clockfly commented:
There are several reference of topologicalOrderIterator in Gearpump, you need to replace them all.

pangolulu commented:
@clockfly Yes, ok.

clockfly commented:
@pangolulu 

How is the result?

pangolulu commented:
@clockfly 

I meet an error, it also happens before. I am debugging it. Maybe it is the problem of gearpump version used in SAMOA.

![error](https://cloud.githubusercontent.com/assets/5796671/12772758/7b122fa4-ca6f-11e5-8f51-ea4e146d6a2f.PNG)

darionyaphet commented:
@pangolulu 

`http://www.drdobbs.com/database/topological-sorting/184410262`  seems wrong  ...

pangolulu commented:
@darionyaphet 

what's the problem?

darionyaphet commented:
@pangolulu the links couldn't point to the wanted page.

pangolulu commented:
@clockfly 

It seems that when I use gearpump ``<version>0.7.5</version>`` in SAMOA ``pom.xml``, it will go wrong. And I change it to ``<version>0.7.4</version>``, it's right now.

I test ``VerticalHoeffdingTree``  with ``topologicalOrderWithCirclesIterator``, the app can be submitted now. But it meets another problem,  ``VerticalHoeffdingTree`` has duplicated edges, so gearpump will raise an exception for duplicated edges (``attribute`` and ``control`` edge).

![vht](https://cloud.githubusercontent.com/assets/5796671/12775408/bdabe738-ca88-11e5-9804-6bd7abf14206.png)

![duplited edges](https://cloud.githubusercontent.com/assets/5796671/12775415/c63e3a0e-ca88-11e5-96cd-0f2938efdc9a.PNG)

Next, I will comment ``require(!dag.hasDuplicatedEdge(), "Graph should not have duplicated edges")``, and see if it will go correct.

clockfly commented:
@huafengw 

About the duplicate edges in SAMOA, I think we took care of it before? By re-organizing the message format?

clockfly commented:
@pangolulu Can you investigate what is wrong on 0.7.5?

huafengw commented:
I check the samoa-gearpump code and we did not combine duplicated edges into one.

pangolulu commented:
@huafengw 

Can we deal with the duplicated edges by a unique edge with different message type?

clockfly commented:
+1
